include/zephyr/drivers/i2c.h

```diff
@@ -179,6 +188,14 @@ typedef int (*i2c_api_target_register_t)(const struct device *dev,
 					struct i2c_target_config *cfg);
 typedef int (*i2c_api_target_unregister_t)(const struct device *dev,
 					  struct i2c_target_config *cfg);
+#ifdef CONFIG_I2C_CALLBACK
+typedef int (*i2c_api_transfer_cb_t)(const struct device *dev,
+				 struct i2c_msg *msgs,
```

Comment:
i still think that it would be better to put all this into a struct. Struct can be later on extended without breaking api (e.g adding list node field if we go into direction of queued transfers).

Current version is suitable only for bus with one device since with more we might hit collision and handling collision is painful. At the end, it should be really fire and forget. Using struct gives path towards that without need of breaking the API in the future.

Reply:
This can be done by providing an asynchronous lock/unlock pair of functions

i2c_lock_async(dev, cb0, data);
i2c_transfer_cb(dev, i2c_msgs, cb1, data);
i2c_transfer_cb(dev, more_i2c_msgs, cb2, data);
i2c_unlock_async(dev, cb3, data);

Obviously not in a row like that but in a callback chain. Callback chaining like this is what RTIO makes look more like a queue of requests and pollable responses.

That's how I was planning on dealing with this issue rather than trying to embed a request queue directly in the device driver. In effect there's an application allocated request queue over all devices rather than each device having its own. This way you can do request chaining or parallelism with most of the I/O task concurrency that something like a promise pipeline might give you.